src/herder/HerderSCPDriver.cpp

```diff
@@ -316,14 +316,23 @@ HerderSCPDriver::validateValueHelper(uint64_t slotIndex, StellarValue const& b,
 
     auto closeTimeOffset = b.closeTime - lastCloseTime;
 
+    // We're in nomination and we ourselves are the leader, so skip block
+    // validation (must already be valid due to block construction) (we already
+    // did signature check earlier in this function, so the value is guaranteed
+    // to be from the local node)
+    bool isAlreadyValidated =
```

Comment:
Could we make this change a bit less error prone and not add this new flag to `checkValid`? Keeping all invariants correct in the future may get pretty hard.

I think that we should just explicitly set the txset validity cache in the right places instead of doing that as a side effect.

Specifically:
* at nomination time (removing the special case here)
* after the call to "trim transactions" (current PR already does that)

In addition, we may want to keep the assert that the txset being cached is valid (when trying to cache its validity) when compiled with "extra-checks" (useful when running tests, etc) as it's an invariant.

I think it's basically the same change, just doesn't risk breaking other (potential future) callers.

Reply:
I've updated herder to explicitly cache the value at nomination, such that we don't need to skip validation here. We still have `txsAreValidated` parameter in checkValid, but it's only used internally withing TxSetFrame, and validation is forced for any external callers (e.g. SCP), so I think this is fairly safe.

Nice idea about the extrachecks invariant, I've added that.